### PR TITLE
Fix for flickering/wrong animation frames in sprites

### LINKED ISSUE #8430: Rendering switching between images with duplicate names in HTML5 build
**Bug description**
For HTML5 applications the animation jumps between images that have the same name but are located under different paths. 

**To Reproduce (REQUIRED)**
1. clone https://github.com/britzl/warbattles-rts
2. Use `Build HTML5` or `Bundle HTML5 application` in editor
3. Press `1` to build a barracks and place it anywhere you like
4. Renderings jumps between different buildings

Please refer to the gif below.

**Expected behavior (REQUIRED)**
I would expect first the `under construction` animation to play and then when ready the idle animation for the barracks.

**Defold version (REQUIRED):**
  - Version 1.6.3
  - Channel editor-alpha
  - Build time 2024-01-19

**Platforms (REQUIRED):**
 - Platforms: HTML5
 - OS: Windows 11
 - Device: desktop PC

**Minimal repro case project (OPTIONAL):**
Easiest is to clone the warbattles-rts repo. Key problem here is images having duplicate names in different paths f.ex `def-war-battles/buildings/barracks/1.png` and `def-war-battles/buildings/command-center/1.png`

**Workaround (OPTIONAL):**
Works with normal build action.

**Screenshots (OPTIONAL):**
bug

### PR DESCRIPTION
**Pull request by JCash**
Atlas animation frames now use internal format "animname/framename" to avoid name collisions

Fixes https://github.com/defold/defold/issues/8430

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AtlasBuilderTest.java [hunk 1]
```diff
@@ -186,4 +192,73 @@ public void testAtlasUtilNoDuplicateFiles2() throws Exception {
         assertTrue(caught);
     }
 
+
+    @Test
+    public void testAtlasFrameIndices() throws Exception {
+        addImage("/a/1.png", 16, 16);
+        addImage("/a/2.png", 16, 16);
+        addImage("/b/1.png", 16, 16);
+        addImage("/b/2.png", 16, 16);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AtlasBuilderTest.java:201]: Adding a unit test is preferable to adding a test project imo.

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/textureset/test/TextureSetGeneratorTest.java [hunk 2]
```diff
@@ -41,67 +43,12 @@ public class TextureSetGeneratorTest {
 
     private static final float EPSILON = 0.000001f;
 
-    public static class MappedAnimDesc extends AnimDesc {
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/textureset/test/TextureSetGeneratorTest.java:44]: Better to use the actual animation iterators in the tests as well

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java [hunk 3]
```diff
@@ -47,33 +47,50 @@
...
+        public List<String> getIds() { // For testing
             return this.ids;
         }
+
+        public List<String> getPaths() {
+            return this.paths;
+        }
+
+        @Override
+        public String getFrameId(int frameIndex) { // local frame index
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java:77]: Opted for an override for simplicity.
In general, I think the animation loading/iteration could be much cleaner, with more specific methods/attributes, and an explicit struct for the animation frames (index, path, id)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java [hunk 4]
```diff
@@ -86,10 +103,10 @@ public AnimDesc nextAnim() {
         }
 
         @Override
-        public Integer nextFrameIndex() {
+        public Integer nextFrameIndex() { // Return the global index of the image that the frame is using
             MappedAnimDesc anim = anims.get(nextAnimIndex - 1);
-            if (nextFrameIndex < anim.getIds().size()) {
-                return imageIds.indexOf(anim.getIds().get(nextFrameIndex++));
+            if (nextFrameIndex < anim.getPaths().size()) {
+                return imagePaths.indexOf(anim.getPaths().get(nextFrameIndex++));
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java:109]: Lookup using the unique project path for the image resource (.png), giving us the index into the image data in the atlas.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java [hunk 5]
```diff
@@ -326,8 +347,9 @@ public static TextureSetResult generateTextureSet(final Project project, IResour
             imageNames.add(imageName);
         }
 
-        List<MappedAnimDesc> animDescs = createAnimDescs(atlas, transformer);;
-        MappedAnimIterator iterator = new MappedAnimIterator(animDescs, imageNames);
+        List<MappedAnimDesc> animDescs = createAnimDescs(atlas, transformer);
+
+        MappedAnimIterator iterator = new MappedAnimIterator(animDescs, imageResourcePaths);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java:352]: We need to use the project paths to resolve the image indices.

File: engine/gamesys/src/gamesys/resources/res_textureset.cpp [hunk 6]
```diff
@@ -103,12 +103,14 @@ namespace dmGameSystem
...
         // This is a mapping from the single frame image names (animation ids) to the frame number
+        const uint32_t* frame_indices = texture_set_ddf->m_FrameIndices.m_Data;
         uint32_t n_image_name_hashes = texture_set_ddf->m_ImageNameHashes.m_Count;
         tile_set->m_FrameIds.SetCapacity(dmMath::Max(1U, (n_image_name_hashes*2)/3), n_image_name_hashes);
         for (uint32_t i = 0; i < n_image_name_hashes; ++i)
         {
             dmhash_t h = texture_set_ddf->m_ImageNameHashes[i];
-            tile_set->m_FrameIds.Put(h, i);
+            uint32_t frame_index = frame_indices[i];
+            tile_set->m_FrameIds.Put(h, frame_index);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_textureset.cpp:113]: We create a mapping from the "animname/framename" -> image resource index.

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AtlasBuilderTest.java [hunk 7]
```diff
@@ -186,4 +192,71 @@ public void testAtlasUtilNoDuplicateFiles2() throws Exception {
...
+        imageNameHashes.add(MurmurHash.hash64("1"));
+        imageNameHashes.add(MurmurHash.hash64("2"));
+        imageNameHashes.add(MurmurHash.hash64("1")); // I'm not sure where the tilecount is 4 / MAWE
+        imageNameHashes.add(MurmurHash.hash64("2"));
+
+        imageNameHashes.add(MurmurHash.hash64("a/1"));
+        imageNameHashes.add(MurmurHash.hash64("a/2"));
+        imageNameHashes.add(MurmurHash.hash64("b/1"));
+        imageNameHashes.add(MurmurHash.hash64("b/2"));
+
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AtlasBuilderTest.java:258]: @vlaaad Here is what the new expected format looks like.

**Review by Jhonnyg — APPROVED** [review]

**Pull request merged**